File: rust/src/dto/symbols.rs
use crate::dto::{FileRangeDto, RangeDto};

/// Symbol kind classification.  Serde `rename_all = "snake_case"` produces
/// values like `"function"`, `"class_"`, `"unknown"` matching Python-side
/// `SymbolKind` StrEnum values exactly.
#[derive(Debug, Clone, PartialEq, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum SymbolKindDto {
    Module,
    #[serde(rename = "class_")]
    Class,
    Function,
    Method,
    Constructor,
    Variable,
    Constant,
    Field,
    Parameter,
    Property,
    TypeParameter,
    #[serde(rename = "import_")]
    Import,
    Unknown,
}

/// A code symbol.  Serialized to Python dict via pythonize, validated
/// by `Symbol.model_validate()` on the Python side.
#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct SymbolDto {
    pub name: String,
    pub qualified_name: Option<String>,
    pub kind: SymbolKindDto,
    pub location: FileRangeDto,
    pub selection_range: Option<RangeDto>,
    pub container_name: Option<String>,
    pub deprecated: bool,
}

File: src/tyo3/models/symbols.py
"""Symbol domain models: symbol discovery operations.

Derived from tyo3-symbols.allium
"""

from __future__ import annotations

from enum import StrEnum

from pydantic import BaseModel, ConfigDict

from tyo3.models.analysis import FileRange, Range

# ── Enums ────────────────────────────────────────────────────────────────


class SymbolKind(StrEnum):
    """Classification of a code symbol."""

    MODULE = "module"
    CLASS = "class_"
    FUNCTION = "function"
    METHOD = "method"
    CONSTRUCTOR = "constructor"
    VARIABLE = "variable"
    CONSTANT = "constant"
    FIELD = "field"
    PARAMETER = "parameter"
    PROPERTY = "property"
    TYPE_PARAMETER = "type_parameter"
    IMPORT = "import_"
    UNKNOWN = "unknown"


# ── Entities ─────────────────────────────────────────────────────────────


class Symbol(BaseModel):
    """A code symbol discovered in a TyO3 project."""

    model_config = ConfigDict(from_attributes=True)

    name: str
    qualified_name: str | None = None
    kind: SymbolKind
    location: FileRange
    selection_range: Range | None = None
    container_name: str | None = None
    deprecated: bool = False
    durable_id: str | None = None
    content_hash: str | None = None
    content_hashes: dict[str, str] = {}


__all__ = [
    "SymbolKind",
    "Symbol",
]
